### PR TITLE
[mono] Exclude unused bundle files in AppleAppBuilder

### PR DESCRIPTION
**Pull request by kotlarmilos**
This PR aims to reduce the bundle size of the HelloiOS app by excluding unused globalization files and the dedup assembly. These files are currently included in the performance measurements but are not actually used by the app.

Here is an example with the unused bundle files included.

SOD - iOS HelloWorld .app Size
Metric                                          |Average           
------------------------------------------------|-------------------
SOD - iOS HelloWorld .app Size                  |29462794.000 bytes |29462794.000 bytes |29462794.000 bytes
pub/HelloiOS.app/Program.dll                    |6144.000 bytes     |6144.000 bytes     |6144.000 bytes
pub/HelloiOS.app/HelloiOS                       |20916632.000 bytes |20916632.000 bytes |20916632.000 bytes
pub/HelloiOS.app/icudt_EFIGS.dat                |550832.000 bytes   |550832.000 bytes   |550832.000 bytes
pub/HelloiOS.app/System.Console.dll             |17408.000 bytes    |17408.000 bytes    |17408.000 bytes
pub/HelloiOS.app/icudt_CJK.dat                  |956416.000 bytes   |956416.000 bytes   |956416.000 bytes
pub/HelloiOS.app/System.Private.CoreLib.aotdata |1825744.000 bytes  |1825744.000 bytes  |1825744.000 bytes
pub/HelloiOS.app/Program.runtimeconfig.json     |1064.000 bytes     |1064.000 bytes     |1064.000 bytes
pub/HelloiOS.app/System.Console.aotdata         |5800.000 bytes     |5800.000 bytes     |5800.000 bytes
pub/HelloiOS.app/System.Private.CoreLib.dll     |1155072.000 bytes  |1155072.000 bytes  |1155072.000 bytes
pub/HelloiOS.app/aot-instances.aotdata          |1033080.000 bytes  |1033080.000 bytes  |1033080.000 bytes
pub/HelloiOS.app/icudt_no_CJK.dat               |1107168.000 bytes  |1107168.000 bytes  |1107168.000 bytes
pub/HelloiOS.app/Program.aotdata                |10944.000 bytes    |10944.000 bytes    |10944.000 bytes
pub/HelloiOS.app/icudt.dat                      |1859648.000 bytes  |1859648.000 bytes  |1859648.000 bytes
pub/HelloiOS.app/Program.deps.json              |12781.000 bytes    |12781.000 bytes    |12781.000 bytes
pub/HelloiOS.app/aot-instances.dll              |3072.000 bytes     |3072.000 bytes     |3072.000 bytes
pub/HelloiOS.app/Info.plist                     |981.000 bytes      |981.000 bytes      |981.000 bytes
pub/HelloiOS.app/aot-instances.cs               |0.000 bytes        |0.000 bytes        |0.000 bytes
pub/HelloiOS.app/PkgInfo                        |8.000 bytes        |8.000 bytes        |8.000 bytes

### CODE REVIEW

**Comment by SamMonoRT** [comment]
@kotlarmilos - how do you determine they are un-used?

**Comment by kotlarmilos** [comment]
> @kotlarmilos - how do you determine they are un-used?

The runtime loads the AOT module of a container assembly using the assembly name (https://github.com/dotnet/runtime/pull/83711), so the actual assembly is not required. As for the globalization files, the HelloiOS sample app should use `icudt.dat`, making the others unnecessary.

I have checked the HelloiOS sample app locally, but can't confirm that other globalization files are never used. @akoeplinger is it safe to remove `cudt_*.dat` from the bundle? Such a change could mainly affect tests/internal bundles.

**Comment by filipnavara** [comment]
> is it safe to remove `cudt_*.dat` from the bundle?

Yes, it is.

File: src/tasks/AppleAppBuilder/Xcode.cs [hunk 1]
```diff
@@ -254,7 +254,7 @@ public void CreateXcodeProject(string projectName, string cmakeDirectoryPath)
         bool useNativeAOTRuntime = false)
     {
         // bundle everything as resources excluding native files
-        var excludes = new List<string> { ".dll.o", ".dll.s", ".dwarf", ".m", ".h", ".a", ".bc", "libmonosgen-2.0.dylib", "libcoreclr.dylib" };
+        var excludes = new List<string> { ".dll.o", ".dll.s", ".dwarf", ".m", ".h", ".a", ".bc", "libmonosgen-2.0.dylib", "libcoreclr.dylib", "aot-instances.dll", "icudt_EFIGS.dat", "icudt_CJK.dat", "icudt_no_CJK.dat" };
```
[the remark this answers is not in the document]
Comment (akoeplinger) [reply src/tasks/AppleAppBuilder/Xcode.cs:257]: alternatively for the icudt* files we could continue filtering them in Xcode.cs, but we should be using a wildcard so we don't need to update the list

[the remark this answers is not in the document]
Comment (kotlarmilos) [reply src/tasks/AppleAppBuilder/Xcode.cs:257]: I completely agree. We can exclude the `icudt_*` files in Xcode.cs. Additionally, I will update AppleAppBuilder to incorporate the use of the `excludes` list, allowing the dedup assembly to be excluded in `ExcludeFromAppDir`.

**Review by ivanpovazan — APPROVED** [review]

File: src/mono/sample/iOS/Program.csproj [hunk 2]
```diff
@@ -109,6 +109,7 @@
         EnableAppSandbox="$(EnableAppSandbox)"
         DiagnosticPorts="$(DiagnosticPorts)"
         StripSymbolTable="$(StripDebugSymbols)"
+        ExcludeFromAppDir="$(_iOSLikeDedupAssembly)"
```
[the remark this answers is not in the document]
Comment (steveisok) [reply src/mono/sample/iOS/Program.csproj:112]: Agreed. I think we could do the same thing to more properties and simplify the appbuilder.

**Review by steveisok — APPROVED** [review]

File: src/mono/sample/iOS/Program.csproj (hunk 2 above)
[the remark this answers is not in the document]
Comment (kotlarmilos) [reply src/mono/sample/iOS/Program.csproj]: Thanks.

> I think we could do the same thing to more properties and simplify the appbuilder.

I suggest we remove redundant properties and simplify the appbuilder in https://github.com/dotnet/runtime/pull/86578.

[the remark this answers is not in the document]
Comment (ivanpovazan) [reply src/mono/sample/iOS/Program.csproj]: > Thanks.
> 
> > I think we could do the same thing to more properties and simplify the appbuilder.
> 
> I suggest we remove redundant properties and simplify the appbuilder in #86578.

Agreed!

**Pull request merged**